### DISCUSSION TITLE
Kill battery notification when state changes

### DISCUSSION THREAD

**Discussion by amtoine** [qtile/qtile#3435] [Ideas]
Let us say that the battery gets low, a notification is thrown!
But when the battery is plugged back, the notification does not go away on next update.

As I let the notifications stay there forever to remind me to plug my battery, they never go away unless killed manually.

**Would be great to add a test that kills the battery popup when the state changes to something that does not require the user to be notified anymore!**

wanted to take a look a it, saw this `id` variable that is returned from `libqtile.utils.send_notification`, but could not find a way to kill the popup from within the `python` code nor from the terminal with `dunstify` for instance :sob:
if possible, it would be amazing to have a few hints and try to open a *pull request* :star_struck:

**Comment by amtoine** [comment]
a workaround I found is to set the `notification_timeout` and the `update_interval` to the same value and remove the check with `self._has_notified` inside `libqtile.widget.battery.Battery.poll()`
this way, the widget tries each update to send a notification. The first time the battery is low a notification is thrown. Next time, as `notification_timeout` and `update_interval` are the same and beacuse `self._has_notified` does not play a role, the previous notification ends naturally and a new one is thrown. When the battery is plugged back in, a next notification won't be thrown.

It is a bit quick and dirty and has a major issue: the notification popup flickrs a bit when the notifications are killed and spawned :thinking:

**Comment by elParaguayo** [comment]
The dbus specification for notifications has a `CloseNotification` method (which takes the message id as an argument) so you'd need to call that.

**Comment by amtoine** [comment]
mm I do not see any `CloseNotification` in the entire source code of `dbus_next`, which is imported and used to send notifications inside `qtile`'s `Battery widget :thinking: 

Maybe i do not understand what you mean by "*dbus specification for notifications*" :laughing:

**Reply by elParaguayo** [reply]
It wouldn't be in `dbus_next`'s code. That is code for interacting with the dbus interface.

You'll need to do something like this;
``` python
import asyncio

from dbus_next.constants import MessageType

from libqtile.utils import _send_dbus_message

async def close_notification(self):
    await _send_dbus_message(
        True,
        MessageType.METHOD_CALL,
        "org.freedesktop.Notifications",
        "org.freedesktop.Notifications",
        "/org/freedesktop/Notifications",
        "CloseNotification",
        "u",
        NOTIFICATION_ID,  # You'll need to have stored this somewhere first
    )

def state_change(self):
    asyncio.create_task(self.close_notification())

```

It's untested but that's what I'd do.

**Reply by amtoine** [reply]
ooooh I get what you were talking about, a CloseNotification that goes inside `_send_dbus_message`, very clear!
i test this :yum:

**Reply by elParaguayo** [reply]
Yes. You need to use dbus because notifications are managed over dbus. So this approach should work whether you use qtile's own notification widget or something like dunst.

**Reply by amtoine** [reply]
gotcha :wink:
